turbopape's Clojure Packages Search from Clojars

turbopape opened the pull request:
**What does your Instant Answer do?**
This IA accesses Clojars.org, the central repository of clojure packages, to search for an artifact and give all possible matches with latest versions and pointers to it on Clojars website.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
This IA solves the problem of getting a description of a Clojure package from Clojars as well as a pointer to it while searching for Clojure artifacts on DuckDuckGo.

**What is the data source for your Instant Answer? (Provide a link if possible)**
https://clojars.org/search?q=some-lib&format=json

**Why did you choose this data source?**
Clojars is the defacto standard repository for clojure packages. Leiningen, for instance, uses it.

**Are there any other alternative (better) data sources?**
As far as know, no. There is cross-clj but it is more or less based on clojars to derive its dependency graphs.

**What are some example queries that trigger this Instant Answer?**
clojure instaparse
clojure lib zip
clojure library something
clojars a-lib
clojure package some-lib
clojure package some lib 
clojars library some lib and some other therms

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
Clojure developers and learners.

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
Yes. to and I A I suggested but I could not claim, as it was mine: 
https://duck.co/ideas/idea/6216/search-for-a-clojure-package-on-clojars

**Which existing Instant Answers will this one supersede/overlap with?**
No one, as far as I know.

**Are you having any problems? Do you need our help with anything?**
no.

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
clojars API is free to use.

**Where did you hear about DuckDuckHack? (For first time contributors)**
I heard about DuckDuckHack after I got interested by DuckDuckGO and adopted it as my search engine. I've red a bit about its open-source architecture, and loved the concept.



**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
![ia-clojars-2](https://cloud.githubusercontent.com/assets/2779167/9290795/fc5e2748-4397-11e5-85bb-cb1563ccc82f.png)

-----
IA Page: https://duck.co/ia/view/clojars

moollaza commented:
@turbopape thanks a lot for this! We'll be sure to review this shortly and give you some feedback!

moollaza commented on lib/DDG/Spice/ClojurePackagesFromClojars.pm:
```diff
@@ -0,0 +1,43 @@
+package DDG::Spice::ClojurePackagesFromClojars;
```
Lets simplify this name! `Clojars` (like Npm, RubyGems, Hoogle) is fine, or `ClojurePackages` (similar to SublimePackages)

moollaza commented on share/spice/clojure_packages_from_clojars/clojure_packages_from_clojars.js:
```diff
@@ -0,0 +1,51 @@
+        // Validate the response (customize for your Spice)
+        var search_terms = DDG.get_query().split(" ");
+
+        function keep_it(candidate) {
```
take a look at using a `relevancy: {}` block in the docs. It'll handle all of this for you :+1:

moollaza commented on share/spice/clojure_packages_from_clojars/clojure_packages_from_clojars.js:
```diff
@@ -0,0 +1,51 @@
+
+                variants: {
+                    tileTitle: 'basic4',
+                    detail: 'light'
```
Don’t think you need this for anything. You're not showing a detail pane :+1:

moollaza commented on share/spice/clojure_packages_from_clojars/clojure_packages_from_clojars.js:
```diff
@@ -0,0 +1,51 @@
+                return {
+                    title: item.group_name + "/" + item.jar_name,
+                    subtitle: item.version,
+                    description: item.description,
```
Seeing as `item.description` exists, you don't need to re-specify it here. We're actually extending the `item` obj with whatever you return here, so you don't lose any of the original properties from `item`

turbopape replied:
Thank you for your feedback. I'll adress all of this and let you know when
I'm finished.
Le 11 août 2015 22:08, "Zaahir Moolla" <notifications@github.com> a écrit :

> In
> share/spice/clojure_packages_from_clojars/clojure_packages_from_clojars.js
> <https://github.com/duckduckgo/zeroclickinfo-spice/pull/2087#discussion_r36799120>
> :
>
> > +            },
> > +            templates: {
> > +                group: 'text',
> > +                detail: false,
> > +                item_detail: false,
> > +
> > +                variants: {
> > +                    tileTitle: 'basic4',
> > +                    detail: 'light'
> > +                }
> > +            },
> > +            normalize: function(item) {
> > +                return {
> > +                    title: item.group_name + "/" + item.jar_name,
> > +                    subtitle: item.version,
> > +                    description: item.description,
>
> Seeing as item.description exists, you don't need to re-specify it here.
> We're actually extending the item obj with whatever you return here, so
> you don't lose any of the original properties from item
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/duckduckgo/zeroclickinfo-spice/pull/2087/files#r36799120>
> .
>

turbopape commented:
@moollaza I pushed a modified version as per your suggestions. But your suggestion about using relevancy is not appliable in my case: actually, I use it to get the terms of the query, and not to filter results. I saw that, for instance, news.js (in the tutorial) used a similar approach than mine.

moollaza commented on share/spice/clojars/clojars.js:
```diff
@@ -0,0 +1,50 @@
+        // Validate the response (customize for your Spice)
+        var search_terms = DDG.get_query().split(" ");
+
+        function keep_it(candidate) {
```
Let's move this out of the callback function, right before `ddg.spice_clojars`

moollaza commented on share/spice/clojars/clojars.js:
```diff
@@ -0,0 +1,50 @@
+        Spice.add({
+            id: "clojars",
+            // Customize these properties
+            name: 'On Clojars',
```
Let's change this to "Software" or "Packages"

moollaza commented on t/Clojars.t:
```diff
@@ -0,0 +1,41 @@
+
+ddg_spice_test(
+    [qw( DDG::Spice::Clojars)],
+    # At a minimum, be sure to include tests for all:
```
nit: you can remove this comment, and the one below. They're just for you to read and see

moollaza commented on lib/DDG/Spice/Clojars.pm:
```diff
@@ -0,0 +1,43 @@
+
+use DDG::Spice;
+
+# Caching - https://duck.co/duckduckhack/spice_advanced_backend#caching-api-responses
```
You can also remove this comment and any others that don't explain this specific code.

moollaza commented on lib/DDG/Spice/Clojars.pm:
```diff
@@ -0,0 +1,43 @@
+spice wrap_jsonp_callback => 1;
+
+
+triggers any => "clojure", "clojars","clojure lib", "clojure library";
```
Let's use `startend` to avoid over triggering. Otherwise we'll be triggering this for any query that has "clojure" anywhere. Most of those queries are likely not to be a search for a packge/module.

moollaza commented on lib/DDG/Spice/Clojars.pm:
```diff
@@ -0,0 +1,43 @@
+    # optional - regex guard
+    # return unless qr/^\w+/;
+
+    return unless $_;    # Guard against "no answer"
```
Do we need to cleanup the query at all? Would user's search something like "clojure package $name" or "clojars package for $name"? I'm not exactly sure how a user might search for a Clojure package, but if those are likely we should handle them appropriately so we don't send extra text to the API.

moollaza commented on lib/DDG/Spice/Clojars.pm:
```diff
@@ -0,0 +1,43 @@
+
+
+spice to => 'https://clojars.org/search?q=$1&format=json&callback={{callback}}';
+spice wrap_jsonp_callback => 1;
```
You don't want `wrap_jsonp_callback` if you're passing a `callback` param to the API. If the API supports the callback param it will wrap it for you. In case it doesn't, then you can use this, and remove `&callback=...` from the `spice to`

moollaza commented on share/spice/clojars/clojars.js:
```diff
@@ -0,0 +1,50 @@
+                item_detail: false,
+
+                variants: {
+                    tileTitle: 'basic4'                  
```
This should be `tile`: 'basic4' for it to work correctly :+1:

turbopape commented:
I addressed all of your remarks.
As per the point regarding cleaning the entry, I will just replace space between terms with hyphens, so if wen enter : clojure incanter cascalog, we'll search incanter-cascalog, as this is the format fo clojure packages naming.

moollaza commented on share/spice/clojars/clojars.js:
```diff
@@ -0,0 +1,51 @@
+                item_detail: false,
+
+                variants: {
+                    tile: 'basic4'                  
```
Instead of `basic4` I think we should use:

```js
tileTitle: '2line',
tileSnippet: 'small'
```

This will reduce the tile height, and give the title 2lines of room so we don't clip it as much.

moollaza commented:
@turbopape thanks for the updates! This is looking great. I think we just need to adjust the snippet size and title size and this should be about done.

@abeyang any thoughts on design? I've suggest we move to a 2line title and small snippet. The result looks like this:

![clojars_packages_for_math_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/9290721/e2b62e30-436a-11e5-8054-7ea9fab86830.png)

turbopape commented:
I changed the design as per your recommendation @moollaza @abeyang.

jagtalon commented on share/spice/clojars/clojars.js:
```diff
@@ -0,0 +1,52 @@
+        // Validate the response (customize for your Spice)
+      
+
+        if (!api_result || api_result.error) {
```
Does `api_result.error` actually show up when there's an error?

turbopape replied:
Right, there's no such thing as api_result.error. Instead, I've put api_result.count, which when shows 0, means that no result was shown.

jagtalon commented on t/Clojars.t:
```diff
@@ -0,0 +1,45 @@
+        caller => 'DDG::Spice::Clojars'
+    ),
+    'clojure lib instaparse' => test_spice(
+    '/js/spice/clojars/instaparse',
```
Nitpick: Please indent this as:

```perl
    'clojure lib instaparse' => test_spice(
        '/js/spice/clojars/instaparse',
        call_type => 'include',
        caller => 'DDG::Spice::Clojars'
    ),
```

turbopape replied:
Done.

jagtalon commented on share/spice/clojars/clojars.js:
```diff
@@ -0,0 +1,52 @@
+                }
+            },
+            normalize: function(item) {
+                return {
```
Check if `group_name`, `jar_name`, and `version` exists before accessing them?

turbopape replied:
I Added the check, I return nothing if any of these fields is not found.

jagtalon commented on lib/DDG/Spice/Clojars.pm:
```diff
@@ -0,0 +1,39 @@
+source "clojars.org";
+icon_url "https://clojars.org/favicon.ico";
+description "This spice accesses clojars.org, the clojure libraries reposisitory, to search for a clojure artifact";
+primary_example_queries "clojure some-lib", "clojars some-lib","clojure lib some-lib","clojure library some-lib";
```
Could you add in real packages in here?

jagtalon replied:
Like probably `clojure template` to show its capabilities.

turbopape replied:
Done.

abeyang commented:
Suggestion is good, @moollaza

jagtalon commented:
Created an IA page for this: https://duck.co/ia/view/clojars

jagtalon commented on share/spice/clojars/clojars.js:
```diff
@@ -0,0 +1,52 @@
+        Spice.add({
+            id: "clojars",
+            // Customize these properties
+            name: 'Packages',
```
Please change the tab name to "Software" like in https://duckduckgo.com/?q=rubygems+sass&ia=software

turbopape replied:
Done.

turbopape commented:
New PR as per all of the remarks.

moollaza commented on share/spice/clojars/clojars.js:
```diff
@@ -0,0 +1,54 @@
+                        url: "https://clojars.org/" + item.group_name + "/" + item.jar_name
+                    };
+                } 
+                return;  
```
this should `return null`

moollaza commented on share/spice/clojars/clojars.js:
```diff
@@ -0,0 +1,54 @@
+        return DDG.isRelevant(candidate, ['clojure', 'clojars', 'lib', 'library','pacakge'], 1, 0);
+    }
+
+        var clojar_artifact = search_terms.filter(keep_it);
```
indent is a little off here

moollaza commented:
@turbopape this LGTM :+1: A couple non-essential fixes to be made. But I'll merge for now :smile:

turbopape commented:
Cool ! (Y)
Le 20 août 2015 20:28, "Zaahir Moolla" <notifications@github.com> a écrit :

> Merged #2087 <https://github.com/duckduckgo/zeroclickinfo-spice/pull/2087>
> .
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/duckduckgo/zeroclickinfo-spice/pull/2087#event-388059730>
> .
>

moollaza commented:
@turbopape congrats! It's live: https://duckduckgo.com/?q=clojure+template&ia=software

Thanks again for taking the time to contribute! We really appreciate it. Feel free to stick around and comment/help on other PR's, tackle open issues or even submit more Instant Answers!
